File: ios-native/Jottery/Jottery/Models/EncryptionMetadata.swift
import Foundation
import GRDB

/// Encryption setup metadata — stored once in the `encryption_metadata` table.
struct EncryptionMetadata: Codable, FetchableRecord, PersistableRecord {
    static let databaseTableName = "encryption_metadata"

    var id: Int = 1
    var salt: String           // Base64-encoded 32-byte salt
    var iterations: Int
    var createdAt: String      // ISO 8601
    var algorithm: String      // "AES-256-GCM"
    var verification: String?  // Encrypted known plaintext for password verification

    // Envelope encryption fields (post-migration)
    var envelopeVersion: Int?       // 1 = envelope active
    var deviceSalt: String?         // Base64-encoded per-device salt
    var localWrappedMaster: String? // JSON {"ciphertext":"...","iv":"..."}
    var wrappingKdfVersion: Int?    // 1 = PBKDF2

    enum CodingKeys: String, CodingKey {
        case id
        case salt
        case iterations
        case createdAt = "created_at"
        case algorithm
        case verification
        case envelopeVersion = "envelope_version"
        case deviceSalt = "device_salt"
        case localWrappedMaster = "local_wrapped_master"
        case wrappingKdfVersion = "wrapping_kdf_version"
    }

    /// The known plaintext used for password verification.
    static let verificationPlaintext = "jottery-vault-ok"

    /// Salt as raw bytes.
    var saltData: Data? {
        Data(base64Encoded: salt)
    }

    /// Create metadata for a new vault.
    static func new(salt: Data, iterations: UInt32, verification: String) -> EncryptionMetadata {
        EncryptionMetadata(
            salt: salt.base64EncodedString(),
            iterations: Int(iterations),
            createdAt: ISO8601DateFormatter.jottery.string(from: Date()),
            algorithm: "AES-256-GCM",
            verification: verification
        )
    }
}

File: android-native/app/src/main/java/com/jottery/android/model/EncryptionMetadata.kt
package com.jottery.android.model

import android.util.Base64
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import com.jottery.android.util.DateUtils

/**
 * Encryption metadata. Singleton table (id=1).
 */
@Entity(tableName = "encryption_metadata")
data class EncryptionMetadata(
    @PrimaryKey
    val id: Int = 1,

    // Base64-encoded 32-byte salt (null when envelope active)
    val salt: String? = null,

    // PBKDF2 iterations (null when envelope active)
    val iterations: Int? = null,

    @ColumnInfo(name = "created_at")
    val createdAt: String,

    val algorithm: String = "AES-256-GCM",

    // Encrypted known plaintext for password verification
    val verification: String? = null,

    // Envelope encryption fields (post-migration)
    @ColumnInfo(name = "envelope_version")
    val envelopeVersion: Int? = null,       // 1 = envelope active

    @ColumnInfo(name = "device_salt")
    val deviceSalt: String? = null,         // Base64-encoded per-device salt

    @ColumnInfo(name = "local_wrapped_master")
    val localWrappedMaster: String? = null,  // JSON {"ciphertext":"...","iv":"..."}

    @ColumnInfo(name = "wrapping_kdf_version")
    val wrappingKdfVersion: Int? = null,     // 1 = PBKDF2
) {
    val saltData: ByteArray?
        get() = try {
            salt?.let { Base64.decode(it, Base64.NO_WRAP) }
        } catch (_: Exception) {
            null
        }

    companion object {
        const val VERIFICATION_PLAINTEXT = "jottery-vault-ok"

        fun new(salt: ByteArray, iterations: Int, verification: String): EncryptionMetadata {
            return EncryptionMetadata(
                salt = Base64.encodeToString(salt, Base64.NO_WRAP),
                iterations = iterations,
                createdAt = DateUtils.nowISO8601(),
                verification = verification,
            )
        }
    }
}
